core/system/src/secure_gateway.c

```diff
@@ -0,0 +1,250 @@
+        HALT_ERROR(NOT_ALLOWED, "The destination (%d) of a secure gateway must be different from its source (%d).",
+                                dst_id, g_active_box);
+    }
+
+    /* Forge a stack frame for the destination box and transfer nargs arguments
+     * from the source stack. */
+    xpsr = ((uint32_t *) src_sp)[7];
+    dst_sp = context_forge_exc_sf(src_sp, dst_id, dst_fn, (uint32_t) secure_gateway_thunk, xpsr, nargs);
+
+    /* Perform the context switch-in to the destination box. */
+    /* This function halts if it finds an error. */
+    context_switch_in(CONTEXT_SWITCH_FUNCTION_GATEWAY, dst_id, src_sp, dst_sp);
+
+    /* Return whether the destination box requires privacy or not. */
+    /* TODO: Context privacy is currently unsupported. */
+    return false;
+}
+
+/** Performs a context switch-out as a result of a secure gateway.
+ *
+ * @internal
+ *
+ * This function is implemented as a wrapper, needed to make sure that the lr
+ * register doesn't get polluted and to provide context privacy during a context
+ * switch. The actual function is ::secure_gateway_context_switch_out. */
+void UVISOR_NAKED secure_gateway_out(uint32_t svc_sp)
+{
+    /* According to the ARM ABI, r0 will have the following value when this
+     * function is called:
+     *   r0 = svc_sp */
+    asm volatile(
+        "push {lr}\n"                                   /* Preserve the lr register. */
+        "bl   secure_gateway_context_switch_out\n"      /* secure_gateway_context_switch_out(svc_sp) */
+        "pop  {lr}\n"                                   /* Restore the lr register. */
+        "pop  {r4-r11}\n"                               /* Restore the callee-saved registers saved at switch-in time. */
+        "bx   lr\n"                                     /* Return. */
+        :: "r" (svc_sp)
+    );
+}
+
+/** Performs a context switch-out from a previous secure gateway.
+ *
+ * @internal
+ *
+ * This function implements ::secure_gateway_out, which is instead only a
+ * wrapper.
+ *
+ * @warning This function trusts the SVCall parameters that are passed to it.
+ *
+ * @param svc_sp[in]    Unprivileged stack pointer at the time of the secure
+ *                      gateway return handler (thunk) */
+void secure_gateway_context_switch_out(uint32_t svc_sp)
+{
+    uint32_t dst_sp;
+    TContextPreviousState * previous_state;
+
+    /* Discard the unneeded exception stack frame from the destination box
+     * stack. The destination box is the currently active one. */
+    dst_sp = context_validate_exc_sf(svc_sp);
+    context_discard_exc_sf(g_active_box, dst_sp);
+
+    /* Perform the context switch back to the previous state.
+     * Upon return, the previous state structure will be populated. */
+    previous_state = context_switch_out(CONTEXT_SWITCH_FUNCTION_GATEWAY);
+
+    /* Copy the return value from the destination stack frame.
+     * Note: We trust the source box stack pointer as stored in the previous
+     *       state.*/
```

Comment:
I hope the stack pointer as stored in the previous state includes our saved `lr` and isn't just the `svc_sp`.